viadot/sources/salesforce.py

```diff
@@ -0,0 +1,192 @@
+        client_id (str): client id to keep the track of API calls.
+        credentials (dict): credentials to connect with. If not provided, will read from local config file.
+        env (Literal): environment information, provides information about credential and connection configuration; defaults to 'DEV'.
+        config_key (str, optional): The key in the viadot config holding relevant credentials. Defaults to None.
+
+    """
+
+    def __init__(
+        self,
+        *args,
+        domain: str = "test",
+        client_id: str = "viadot",
+        credentials: Dict[str, Any] = None,
+        env: Literal["DEV", "QA", "PROD"] = "DEV",
+        config_key: str = None,
+        **kwargs,
+    ):
+
+        credentials = credentials or get_source_credentials(config_key) or {}
+
+        if credentials is None:
+            raise CredentialError("Please specify the credentials.")
+
+        super().__init__(*args, credentials=credentials, **kwargs)
+
+        if env.upper() == "DEV":
+            self.salesforce = SF(
+                username=self.credentials.get("username"),
+                password=self.credentials.get("password"),
+                security_token=self.credentials.get("token"),
+                domain=domain,
+                client_id=client_id,
+            )
+        elif env.upper() == "QA":
+            self.salesforce = SF(
+                username=self.credentials.get("username"),
+                password=self.credentials.get("password"),
+                security_token=self.credentials.get("token"),
+                domain=domain,
+                client_id=client_id,
+            )
+        elif env.upper() == "PROD":
+            self.salesforce = SF(
+                username=self.credentials.get("username"),
+                password=self.credentials.get("password"),
+                security_token=self.credentials.get("token"),
+                domain=domain,
+                client_id=client_id,
+            )
+        else:
+            raise ValueError("The only available environments are DEV, QA, and PROD.")
+
+    def upsert(
+        self,
+        df: pd.DataFrame,
+        table: str,
+        external_id: str = None,
+        raise_on_error: bool = False,
+    ) -> None:
+
+        if df.empty:
+            self.logger.info("No data to upsert.")
+            return
+
+        if external_id and external_id not in df.columns:
+            raise ValueError(
+                f"Passed DataFrame does not contain column '{external_id}'."
+            )
+
+        table_to_upsert = getattr(self.salesforce, table)
+        records = df.to_dict("records")
+        records_cp = records.copy()
+
+        for record in records_cp:
+            response = 0
+            if external_id:
+                if record[external_id] is None:
+                    continue
```

Comment:
This begs a comment

Reply:
[🎨 Improved retrieving externalID](https://github.com/dyvenia/viadot/pull/629/commits/97e6dbb9a79abf04b0787c07e8f44830a654fb8f)

Remove test code:

[⏪ Removed tests code](https://github.com/dyvenia/viadot/pull/629/commits/5625cee1776787cbea2faf7ab6a79067fb7491e5)